cocos/core/renderer/scene/submodel.ts

```diff
@@ -70,9 +84,72 @@ export class SubModel {
             this._inputAssembler.destroy();
             this._inputAssembler = null;
         }
-        for (let i = 0; i < this.passes.length; i++) {
-            this.passes[i].destroyPipelineState(this._psos![i]);
-        }
+        this.destroyPipelineRelatedResource();
         this._material = null;
     }
+
+    public update() {
+        for (let i = 0; i < this.passes.length; ++i) {
+            const pass = this.passes[i];
+            pass.update();
+            this._psoCreateInfos[i].hash = pass.hash;
+        }
+
+        this.updateLayout();
+    }
+
+    public updateLayout() {
+        for (let i = 0; i < this._psoCreateInfos.length; ++i) {
+            let psoCreateInfo = this._psoCreateInfos[i];
+            if (psoCreateInfo) {
+                psoCreateInfo.bindingLayout.update();
+            }
+        }
+    }
+
+    public onPipelineStateChanged() {
+        const materail = this._material;
+        if (!materail) {
+            return;
+        }
+
+        const passes = materail.passes;
+        for (let j = 0; j < passes.length; ++j) {
+            const pass = passes[j];
+            pass.beginChangeStatesSilently();
+            pass.tryCompile(); // force update shaders
+            pass.endChangeStatesSilently();
+        }
+        this._psoCreateInfos = [];
```

Comment:
`this._psoCreateInfos.length = 0` should be enough

Reply:
Ok.